### PR TITLE
feat: support recovering parts from the mempool

### PR DESCRIPTION
**Pull request by rach-id**
## Description
 
Closes https://github.com/celestiaorg/celestia-core/issues/1661

#### PR checklist

- [ ] Tests written/updated
- [ ] Changelog entry added in `.changelog` (we use
  [unclog](https://github.com/informalsystems/unclog) to manage our changelog)
- [ ] Updated relevant documentation (`docs/` or `spec/`) and code comments

### CODE REVIEW

File: consensus/propagation/types/unmarshalling.go [hunk 1]
```diff
@@ -0,0 +1,100 @@
...
+}
+
+// UnmarshalledTx is an intermediary type that allows keeping the transaction metadata,
+// its Key and the actual tx bytes.
+// This will be used to create the parts from the local txs.
+type UnmarshalledTx struct {
+	MetaData TxMetaData
+	Key      types.TxKey
+	TxBytes  []byte
+}
```
Comment (evan-forbes) [inline consensus/propagation/types/unmarshalling.go:100]: in the future, we might want to consider use this type as `types.Tx`, as that simplifies the caching of metadata and hashes

File: consensus/propagation/types/unmarshalling_test.go [hunk 2]
```diff
@@ -0,0 +1,90 @@
...
+				End:   uint32(pos.End),
+				Hash:  block.Txs[i].Hash(),
+			},
+			Key:     txKey,
+			TxBytes: marshalledTx,
+		}
+	}
+
+	// generate all the possible combinations for the provided number of transactions
+	txsCombinations := GenerateTxsCombinations(numberOfTxs)
```
Comment (evan-forbes) [inline consensus/propagation/types/unmarshalling_test.go:54]: nice lol

File: consensus/propagation/types/unmarshalling_test.go [hunk 3]
```diff
@@ -0,0 +1,90 @@
...
+	"github.com/stretchr/testify/require"
+	cmtrand "github.com/tendermint/tendermint/libs/rand"
+	"github.com/tendermint/tendermint/proto/tendermint/mempool"
+	"github.com/tendermint/tendermint/state"
+	"github.com/tendermint/tendermint/types"
+)
+
+// TestTxsToParts extensive testing of the txs to parts method
+// that recovers the parts from mempool txs.
+func TestTxsToParts(t *testing.T) {
```
Comment (evan-forbes) [inline consensus/propagation/types/unmarshalling_test.go:19]: [could be a followup issue]
I like this test and have written similar style heavy weight random tests before. I think this is combining correctness testing and fuzzing however. It might actually be better if we have table driven tests for as many edge cases as we can think of, then have a fuzzer ontop

**Review by evan-forbes — COMMENTED** [review]
nice! only one or two blocking comments

File: consensus/propagation/commitment.go [hunk 4]
```diff
@@ -88,6 +90,65 @@ func (blockProp *Reactor) handleCompactBlock(cb *proptypes.CompactBlock, peer p2
...
+		marshalledTx, err := proto.Marshal(&protoTxs)
+		if err != nil {
+			blockProp.Logger.Error("failed to encode tx", "err", err, "tx", txMetaData)
+			continue
+		}
+
+		txsFound = append(txsFound, proptypes.UnmarshalledTx{MetaData: txMetaData, Key: txKey, TxBytes: marshalledTx})
+	}
+	if len(txsFound) == 0 {
+		// no compact block transaction was found locally
```
[the remark this answers is not in the document]
Comment (rach-id) [reply consensus/propagation/commitment.go:138]: actually I wrote them 🤣 🤣

File: consensus/propagation/types/unmarshalling_test.go [hunk 5]
```diff
@@ -0,0 +1,90 @@
...
+	cleanup, _, sm := state.SetupTestCase(t)
+	t.Cleanup(func() {
+		cleanup(t)
+	})
+	numberOfTxs := 18 // increasing the number of transactions increases the test time exponentially
+	txs := make([]types.Tx, 0, numberOfTxs)
+	for i := 0; i < numberOfTxs; i++ {
+		txs = append(txs, cmtrand.Bytes(int(types.BlockPartSizeBytes/3)))
+	}
```
[the remark this answers is not in the document]
Comment (rach-id) [reply consensus/propagation/types/unmarshalling_test.go:28]: yes, we could test for larger, I'll add another test case. The thinking was that since for the algorithm its just a bunch of bytes, generating small transactions will help more cases

[the remark this answers is not in the document]
Comment (rach-id) [reply consensus/propagation/types/unmarshalling_test.go]: 44ed0f4ddedce76263f0cd4d742855521e45fc13

File: consensus/propagation/types/unmarshalling_test.go (hunk 3 above)
Comment (rach-id) [reply consensus/propagation/types/unmarshalling_test.go:19]: https://github.com/celestiaorg/celestia-core/issues/1666

File: consensus/propagation/types/unmarshalling_test.go (hunk 5 above)
[the remark this answers is not in the document]
Comment (evan-forbes) [reply consensus/propagation/types/unmarshalling_test.go]: 3 is good
